### PR TITLE
add BlasWrapper

### PR DESCRIPTION
**Pull request by zhangxiaoli73**
## What changes were proposed in this pull request?

wrapper blas layer to dnn layer, then we can use blas layer in dnn graph

## How was this patch tested?

unit tests

## Related links or issues (optional)
fixed https://github.com/intel-analytics/BigDL/issues/XXX

### CODE REVIEW

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala [hunk 1]
```diff
@@ -0,0 +1,125 @@
...
+
+import com.intel.analytics.bigdl.Module
+import com.intel.analytics.bigdl.mkl.Memory
+import com.intel.analytics.bigdl.nn.abstractnn.{AbstractModule, Activity, TensorModule}
+import com.intel.analytics.bigdl.tensor.Tensor
+import com.intel.analytics.bigdl.tensor.TensorNumericMath.TensorNumeric
+import com.intel.analytics.bigdl.utils.Shape
+import spire.syntax.module
+
+// only support wrap bigdl to support nchw or nc
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala]: what else that we cannot support ？

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala [hunk 2]
```diff
@@ -0,0 +1,125 @@
...
+private[bigdl] class BlasWrapper(val module: TensorModule[Float]) extends MklDnnLayer {
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala]: private val module ?

Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala]: and why it's hard-coded to float ? what if int or fp16?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala (hunk 1 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala]: Now only support 2 or 4 dims inputs. Other cases will be error.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala (hunk 2 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala]: Now BlasWrapper is only used when converting to dnn graph. As tests are going on, I think not expose BlasWrapper  to users is better.

Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala]: As for  hard-coded to float, dnn layer now only supports float tensor, so here tensor data type should be float.
For int or fp16 type, I am not sure how to define the model.

Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala]: I mean the parameter `module`  should be private, right ?

Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala]: In future, we may have in8, fp16 training/inference, so probably we need to support them, just make it generic ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+private[bigdl] class BlasWrapper(val module: AbstractModule[Activity, Activity, Float])
+  extends MklDnnLayer {
+
+  require(!module.isInstanceOf[MklDnnModule], "Only support wrapper blas layer to dnn layer")
+
+  output = module.output
+  gradInput = module.gradInput
+
+  private def inferFormats(inputs: Array[MemoryData]): Int = {
+    // remind: here assume all shapes in inputs should be same
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala]: reminder

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala [hunk 4]
```diff
@@ -0,0 +1,139 @@
...
+
+    inputs(0).layout match {
+      case Memory.Format.nhwc => Memory.Format.nhwc
+      case Memory.Format.nc => Memory.Format.nc
+      case _ => Memory.Format.nchw
+    }
+  }
+
+  override private[mkldnn] def initFwdPrimitives(inputs: Array[MemoryData], phase: Phase) = {
+    // todo: only support model having implemented computeOutputShape
```
[the remark this answers is not in the document]
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala]: sorry, here should not be "todo", but "reminder". Only having implemented computeOutputShape blas layer can wrapped to dnn layer.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala [hunk 5]
```diff
@@ -0,0 +1,139 @@
...
+    val inputShape = inputs.map(in => Shape(in.shape))
+    val outputShape = if (inputShape.length == 1) {
+      // single shape
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala]: Do we need this comment ?  the if condition will tell us obviously right ?

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala [hunk 6]
```diff
@@ -0,0 +1,139 @@
...
+      List(module.computeOutputShape(inputShape(0)))
+    } else {
+      // multi shape
+      val out = module.computeOutputShape(MultiShape(inputShape.toList))
+      if (out.isInstanceOf[MultiShape]) out.toMulti() else List(out)
+    }
+    val outDim = outputShape(0).toSingle().length
+    require(outDim == 4 || outDim == 2,
+      s"just support output shape dim is 2 or 4, but get ${outDim}")
```
Comment (wzhongyuan) [inline spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala]: Only shap dim 2 and 4 supported ?

**Comment by wzhongyuan** [comment]
Only some small comments, other looks good to me

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala (hunk 6 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala]: yeah, dim 3 has no corresponding memory format in mkl-dnn.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala (hunk 5 above)
Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala]: ok, just a reminder.

File: spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala (hunk 6 above)
Comment (wzhongyuan) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala]: I mean  make a slight change to the description to make it aligned with other exception

Comment (zhangxiaoli73) [reply spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/mkldnn/BlasWrapper.scala]: ok, changed

**Comment by zhangxiaoli73** [comment]
Jenkins pass:
http://mlt-bddl01.sh.intel.com:8080/job/BigDL-PRVN-lite/301/

**Pull request merged**